src/types/function.rs

```diff
@@ -39,6 +76,38 @@ impl PyCFunction {
         )
     }
 
+    /// Create a new function from a closure.
+    pub fn new_closure<F>(f: F, py: Python) -> PyResult<&PyCFunction>
+    where
+        F: Fn(&types::PyTuple, Option<&types::PyDict>) -> PyResult<PyObject>,
+    {
+        let boxed_fn: Box<dyn Fn(&types::PyTuple, Option<&types::PyDict>) -> PyResult<PyObject>> =
+            Box::new(f);
+        let function_ptr = Box::into_raw(Box::new(boxed_fn));
+        let capsule_ptr = unsafe {
+            ffi::PyCapsule_New(
+                function_ptr as *mut c_void,
+                closure_capsule_name(),
+                Some(drop_closure),
+            )
+        };
+        let method_def = methods::PyMethodDef::cfunction_with_keywords(
+            "pyo3_closure",
```

Comment:
Would it be useful to allow this to be passed as an argument (still `&'static str` of course)?

Reply:
I've exposed it, as it may indeed be useful. Longer term, being able to pass an owned string would be even better, but no need to cram everything in here.

Reply:
Hm, do we even want to name anonymous functions? I don't have a strong opinion either way. However given that you can't (ergonomically) create `&'static str` at runtime, you can't easily create names dynamically at runtime.

Reply:
I don't have a strong opinion here, the names appear when printing the function object itself, e.g.:
```
<built-in method my-function-name of PyCapsule object at 0x7fb400b55660>
```
But not sure whether that makes it worth it to expose to the end user. @birkenfeld any thoughts there?

Reply:
How does it look in tracebacks? That's where I was initially thinking it might be useful.

Reply:
I actually don't see the function name (or the closure at all) appearing in the traceback though I might be doing something wrong here.
```rust
      let counter_fn = move |_args: &types::PyTuple, _kwargs: Option<&types::PyDict>| {
          let gil = Python::acquire_gil();
          let py = gil.python();
          py.run("raise Exception('banana')", None, None)?;
          Ok(())
      };
```
This prints the following traceback:
```python
import traceback
try:
  counter_py()
except Exception:
  print(traceback.format_exc())

>> Traceback (most recent call last):
>>  File "<string>", line 3, in <module>
>>  File "<string>", line 1, in <module>
>> Exception: banana
```

Reply:
ok, then I guess the name is not useful.

Reply:
Ok, I've switched it back to being hard-coded then.